Add basic role and integration for Grafana Loki

sysvinit opened the pull request:
This change introduces minimal, bare-bones platform integration for Grafana Loki, intended to serve as a basis for further development.

This includes:
- A role which can be assigned to VMs, which installs and provisions an instance of Loki. Currently this is statically configured to use filesystem-backed storage.
- Automatic configuration of promtail, a log-shipping client for Loki, when a Loki server is discovered in the same resource group. Currently only a single Loki server is supported. promtail is configured to read the system journal in JSON format, as this reads annotations and metadata which would otherwise be lost if reading only the plain journal messages. The Loki query language LogQL supports decoding JSON log messages with the `json` filter.
- Initial integration with Grafana in the existing statshost role. Hosts with the statshost-master role for a per-RG Grafana instance which also have the loki role enabled will have a data source added to allow Grafana to read and query data from the local Loki instance.

@flyingcircusio/release-managers

## Release process

Impact: internal. 

Changelog: none. This role is still under development, and is not yet intended for public use.

### PR release workflow (internal)

- [x] PR has internal ticket
- [x] internal issue ID (PL-…) part of branch name
- [x] internal issue ID mentioned in PR description text
- [x] ticket is on Platform agile board
- [x] ticket state set to *Pull request ready*
- [x] if ticket is more urgent than within the next few days, directly contact a member of the Platform team

## Design notes

- [x] Provide a feature toggle if the change might need to be adjusted/reverted quickly depending on context. Consider whether the default should be `on` or `off`. Example: rate limiting.
  - The log retention time is intentionally configurable, as this may differ between use cases and applications.
- [x] All customer-facing features and (NixOS) options need to be discoverable from documentation. Add or update relevant documentation such that hosted and guided customers can understand it as well.
  - No documentation yet, this is a development preview which is subject to change.

## Security implications

- [x] [Security requirements](https://wiki.flyingcircus.io/System_Development_Guideline#Security_requirement_principles_and_testing) defined? (WHERE)

This is a basic first implementation, we're reviewing and considering overall security measures in PL-133025.

At this point we're limiting retention to 30 days by default (overridable per loki instance). We are relying on the SRV firewall as usual and find unauthed/cleartext submission acceptable for now.

To reduce impact if attackers have access on a non-loki machine, we're limiting access to the loki API to selected endpoints.

- [x] Security requirements tested? (EVIDENCE)
  - Integrations tested incrementally in a development environment to ensure correct functioning.
  - [x] data separation is handled with regular dev/whq/rzob and also RG separation for customer involvement
  - [x] attack surface area minimisation: no auth used, but covered by RG firewalling, service only on SRV
  - [x] loki runs as separate user

sysvinit committed:
loki: add basic role

Simple configuration with static filesystem configuration.

PL-132981

sysvinit committed:
promtail: add basic log shipping configuration

This automatically enables promtail when a Loki server is present in
the same resource group. Currently only a single Loki server is
supported.

PL-132981

sysvinit committed:
statshost: add loki data source to grafana configuration

If the loki role is also enabled on the same host as statshost-master,
then add a Loki data source to the Grafana configuration.

PL-132981

sysvinit committed:
loki: add options for configuring S3 storage

Provide a minimal set of options for providing configuration for an S3
endpoint used for storing log data, and for providing the required
storage schedule entries in order to use the configured S3 endpoint.

PL-132891

ctheune reviewed (commented):
That looks like a great simple start.

sysvinit committed:
loki: use nginx as reverse proxy for access control

Limit access to the Loki API from external machines to only the data
ingestion and querying endpoints. Configure Grafana to communciate
with the Loki API over localhost without restrictions.

PL-132981
